processing/src/main/java/org/apache/druid/query/operator/NaivePartitioningOperator.java

```diff
@@ -110,28 +79,7 @@ public Closeable goOrContinue(Closeable continuation, Receiver receiver)
           @Override
           public Signal push(RowsAndColumns rac)
```

Comment:
make the `Receiver` a static inner class; it should have the `Iterator` as its field.

handlePush naturally wants to be a method of it

* the full `if (cont.iter != null) ` part should also go into the abstract;
* put the full body of the `while` into a method in the abstract
* the glueing should override that method and before calling the `super()` it could check if there are more elements and save that and use that when the `Receiver` is built

Reply:
A static receiver wouldn't have access to methods like `ensureMaxRowsMaterializedConstraint()` which are also called from non-static methods.

Also, can you elaborate why should we make receiver static? Even if we had a StaticReceiver, we would still have to do `new StaticReceiver()` with the required state.

For example, to add a static receiver in NaivePartitioningOperator, I added the following `StaticReceiver` class and then replaced `new Receiver() {}` with `new StaticReceiver(receiver, iterHolder)`:

```java
static class StaticReceiver implements Receiver
  {
    private final Receiver delegate;
    private final AtomicReference<Iterator<RowsAndColumns>> iterHolder;

    public StaticReceiver(
        Receiver delegate,
        AtomicReference<Iterator<RowsAndColumns>> iterHolder
    ) {
      this.delegate = delegate;
      this.iterHolder = iterHolder;
    }

    @Override
    public Signal push(RowsAndColumns rac)
    {
      return handlePush(rac, delegate, iterHolder);
    }

    @Override
    public void completed()
    {
      if (iterHolder.get() == null) {
        delegate.completed();
      }
    }
  }
```


I think we're not on the same page wrt how the static receiver needs to look like. Can you share your thoughts on the above?

Reply:
doesn't seem like a dealbreaker...as a last resort  `ensureMaxRowsMaterializedConstraint` can be static and  pass the 2 integers for it?

Reply:
@kgyrtkirk The iterator needs to be created in the receiver's push() method.

Having a static receiver doesn't let us create the iterator there, since getIteratorForRAC() is not static.

getIteratorForRAC() can't be made static since it's supposed to be an abstract method, with different implementations in NaivePartitioningOperator and GlueingPartitioningOperator.

If I remove getIteratorForRAC() from the abstract class, and just have it as a regular non-overridden static method, then how to pass the iterator to the static receiver? I can't pass it in like `super.push(rac, iteratorForRac)` since it violates the method signature of `push()`.

Also, trying to make everything static also ends up requiring us to pass unnecessary variables as method params, and structure everything just to make the receiver static, which seems unnecessary and not clean to me. We anyway would still have to do `new StaticReceiver(field1, field2,...)` everytime instead of `new Receiver()`.

Thoughts?

Reply:
Have made local changes to make receiver and the iterator class static. The code doesn't seem clean though, as I had to pass a bunch of params. Also trying to see what can be moved to Abstract class with those changes. (This is all on my local right now, just updating this info here)

Reply:
@kgyrtkirk Have pushed the change to the PR to make the relevant classes static. Appreciate if you could take a look again, thanks!